File: python/src/dexcost/integrations/langchain.py
"""LangChain-compatible callback handler for dexcost cost tracking.

Implements US-032: a callback handler that captures LLM calls made through
LangChain and records them as dexcost events.  Uses duck typing so that
the ``langchain`` package is **not** required at import time.

Usage::

    from dexcost.integrations import DexcostCallbackHandler
    from dexcost.tracker import CostTracker

    tracker = CostTracker()
    handler = DexcostCallbackHandler(tracker)

    # Pass to LangChain as a callback:
    llm = ChatOpenAI(callbacks=[handler])
"""

from __future__ import annotations

import logging
import time
import uuid
from decimal import Decimal
from typing import Any

from dexcost.context import get_current_task
from dexcost.models.event import Event
from dexcost.pricing import PricingEngine
from dexcost.storage.protocol import StorageBackend

_log = logging.getLogger(__name__)


class DexcostCallbackHandler:
    """LangChain-compatible callback handler for cost tracking.

    This handler implements the same method signatures that LangChain's
    ``BaseCallbackHandler`` uses, but does **not** inherit from it.  This
    allows the handler to work when LangChain is installed (duck typing)
    without requiring LangChain as a dependency.

    Args:
        tracker: A :class:`~dexcost.tracker.CostTracker` instance.  The
            handler uses its storage backend and pricing engine.
    """

    def __init__(self, tracker: Any) -> None:
        self._storage: StorageBackend = tracker._storage
        self._pricing: PricingEngine = tracker._pricing
        self._pending: dict[str, dict[str, Any]] = {}

    # ------------------------------------------------------------------
    # LangChain callback interface
    # ------------------------------------------------------------------

    def on_llm_start(
        self,
        serialized: dict[str, Any],
        prompts: list[str],
        *,
        run_id: uuid.UUID | None = None,
        **kwargs: Any,
    ) -> None:
        """Called when an LLM starts generating.

        Records the start time and extracts the model name from the
        serialized kwargs so they are available when the call completes.
        """
        key = str(run_id) if run_id is not None else str(uuid.uuid4())
        model = _extract_model(serialized, kwargs)
        self._pending[key] = {
            "start_time": time.perf_counter(),
            "model": model,
        }

    def on_llm_end(
        self,
        response: Any,
        *,
        run_id: uuid.UUID | None = None,
        **kwargs: Any,
    ) -> None:
        """Called when an LLM finishes generating.

        Extracts token usage from the response, computes cost via the
        pricing engine, creates an :class:`Event`, and inserts it into
        storage if a task context is active.
        """
        task = get_current_task()
        if task is None:
            _log.warning(
                "DexcostCallbackHandler.on_llm_end called outside a task "
                "context — skipping event recording."
            )
            self._cleanup_pending(run_id)
            return

        key = str(run_id) if run_id is not None else None
        pending = self._pending.pop(key, {}) if key else {}
        model = pending.get("model", "unknown")
        start_time = pending.get("start_time")
        latency_ms: int | None = None
        if start_time is not None:
            latency_ms = int((time.perf_counter() - start_time) * 1000)

        # Extract token usage from the LLMResult
        input_tokens, output_tokens = _extract_tokens(response)
        has_usage = input_tokens > 0 or output_tokens > 0

        if has_usage:
            try:
                cost_result = self._pricing.get_cost(model, input_tokens, output_tokens)
            except Exception:
                cost_result = None
            if cost_result is not None:
                cost_usd = cost_result.cost_usd
                cost_confidence = "computed"
                pricing_source = cost_result.pricing_source
                pricing_version = cost_result.pricing_version
            else:
                cost_usd = Decimal("0")
                cost_confidence = "unknown"
                pricing_source = "unknown"
                pricing_version = None
        else:
            cost_usd = Decimal("0")
            cost_confidence = "unknown"
            pricing_source = "unknown"
            pricing_version = None

        event = Event(
            task_id=task.task_id,
            event_type="llm_call",
            cost_usd=cost_usd,
            cost_confidence=cost_confidence,
            pricing_source=pricing_source,
            pricing_version=pricing_version,
            provider="langchain",
            model=model,
            input_tokens=input_tokens if has_usage else None,
            output_tokens=output_tokens if has_usage else None,
            latency_ms=latency_ms,
        )
        self._storage.insert_event(event)

    def on_llm_error(
        self,
        error: BaseException,
        *,
        run_id: uuid.UUID | None = None,
        **kwargs: Any,
    ) -> None:
        """Called when an LLM call errors.

        Records a failure event if a task context is active.  Never raises.
        """
        task = get_current_task()
        if task is None:
            _log.warning(
                "DexcostCallbackHandler.on_llm_error called outside a task "
                "context — skipping event recording."
            )
            self._cleanup_pending(run_id)
            return

        key = str(run_id) if run_id is not None else None
        pending = self._pending.pop(key, {}) if key else {}
        model = pending.get("model", "unknown")
        start_time = pending.get("start_time")
        latency_ms: int | None = None
        if start_time is not None:
            latency_ms = int((time.perf_counter() - start_time) * 1000)

        error_type = type(error).__name__
        event = Event(
            task_id=task.task_id,
            event_type="llm_call",
            cost_usd=Decimal("0"),
            cost_confidence="unknown",
            pricing_source="unknown",
            provider="langchain",
            model=model,
            latency_ms=latency_ms,
            details={"error": str(error), "error_type": error_type},
        )
        self._storage.insert_event(event)

    # ------------------------------------------------------------------
    # Internal helpers
    # ------------------------------------------------------------------

    def _cleanup_pending(self, run_id: uuid.UUID | None) -> None:
        """Remove pending state for a run without recording an event."""
        key = str(run_id) if run_id is not None else None
        if key:
            self._pending.pop(key, None)


def _extract_model(serialized: dict[str, Any], kwargs: Any) -> str:
    """Extract the model name from LangChain's serialized dict.

    LangChain passes the LLM configuration in ``serialized["kwargs"]``.
    The model name can appear under ``"model"``, ``"model_name"``, or
    ``"model_id"``.
    """
    kw = serialized.get("kwargs", {})
    for key in ("model", "model_name", "model_id"):
        val = kw.get(key)
        if val:
            return str(val)
    # Fallback: check top-level kwargs
    for key in ("model", "model_name", "model_id"):
        val = kwargs.get(key) if isinstance(kwargs, dict) else None
        if val:
            return str(val)
    return "unknown"


def _extract_tokens(response: Any) -> tuple[int, int]:
    """Extract token counts from a LangChain LLMResult.

    LangChain stores usage in ``response.llm_output["token_usage"]`` with
    keys ``prompt_tokens`` and ``completion_tokens``.
    """
    llm_output = getattr(response, "llm_output", None)
    if llm_output is None:
        # Also try dict access for mock objects
        if isinstance(response, dict):
            llm_output = response.get("llm_output")
        if llm_output is None:
            return 0, 0

    if isinstance(llm_output, dict):
        token_usage = llm_output.get("token_usage", {})
    else:
        token_usage = getattr(llm_output, "token_usage", {}) or {}

    if not isinstance(token_usage, dict):
        return 0, 0

    try:
        input_tokens = int(token_usage.get("prompt_tokens", 0) or 0)
        output_tokens = int(token_usage.get("completion_tokens", 0) or 0)
    except (ValueError, TypeError):
        input_tokens = 0
        output_tokens = 0
    return input_tokens, output_tokens

File: rust/src/integrations/langchain.rs
//! LangChain-compatible callback handler for dexcost cost tracking.
//!
//! Rust has no canonical LangChain library, so this module provides a
//! framework-agnostic equivalent of the Python SDK's `DexcostCallbackHandler`
//! (`integrations/langchain.py`). A caller wiring dexcost into a LangChain-like
//! pipeline drives the handler through the same lifecycle:
//!
//! 1. [`DexcostCallbackHandler::on_llm_start`] — records the start time and the
//!    model name for a run.
//! 2. [`DexcostCallbackHandler::on_llm_end`] — extracts token usage, computes
//!    cost via the [`PricingEngine`], and records an `llm_call` event.
//! 3. [`DexcostCallbackHandler::on_llm_error`] — records a failure `llm_call`
//!    event carrying `error_type` in its details.
//!
//! Events are attributed to the supplied `task_id` and appended to the shared
//! [`EventBuffer`].

use std::collections::HashMap;
use std::sync::Arc;
use std::time::Instant;

use tokio::sync::Mutex;

use crate::core::models::{CostConfidence, CostEvent, EventType, PricingSource};
use crate::pricing::engine::PricingEngine;
use crate::transport::buffer::EventBuffer;

/// Per-run state captured between `on_llm_start` and `on_llm_end`/`on_llm_error`.
#[derive(Debug, Clone)]
struct PendingRun {
    start_time: Instant,
    model: String,
}

/// LangChain-compatible callback handler that records LLM calls as dexcost
/// cost events.
///
/// Mirrors Python's `DexcostCallbackHandler`: it does not depend on any
/// LangChain crate and is driven entirely through its public methods.
pub struct DexcostCallbackHandler {
    buffer: Arc<Mutex<EventBuffer>>,
    pricing: Arc<Mutex<PricingEngine>>,
    /// Task that recorded events are attributed to.
    task_id: String,
    /// In-flight runs keyed by `run_id`.
    pending: Mutex<HashMap<String, PendingRun>>,
}

impl DexcostCallbackHandler {
    /// Creates a new handler. All events are attributed to `task_id`.
    pub fn new(
        buffer: Arc<Mutex<EventBuffer>>,
        pricing: Arc<Mutex<PricingEngine>>,
        task_id: impl Into<String>,
    ) -> Self {
        Self {
            buffer,
            pricing,
            task_id: task_id.into(),
            pending: Mutex::new(HashMap::new()),
        }
    }

    /// Called when an LLM starts generating. Records the start time and model
    /// name so they are available when the run completes.
    pub async fn on_llm_start(&self, run_id: &str, model: &str) {
        let mut pending = self.pending.lock().await;
        pending.insert(
            run_id.to_string(),
            PendingRun {
                start_time: Instant::now(),
                model: model.to_string(),
            },
        );
    }

    /// Called when an LLM finishes generating. Computes cost from the token
    /// usage and records an `llm_call` event. Returns the recorded event.
    pub async fn on_llm_end(
        &self,
        run_id: &str,
        input_tokens: i64,
        output_tokens: i64,
    ) -> CostEvent {
        let pending = self.pending.lock().await.remove(run_id);
        let (model, latency_ms) = match pending {
            Some(p) => (
                p.model,
                Some(p.start_time.elapsed().as_millis() as i64),
            ),
            None => ("unknown".to_string(), None),
        };

        let has_usage = input_tokens > 0 || output_tokens > 0;

        let mut event = CostEvent::new(&self.task_id, EventType::LlmCall);
        event.provider = Some("langchain".to_string());
        event.model = Some(model.clone());
        event.latency_ms = latency_ms;

        if has_usage {
            event.input_tokens = Some(input_tokens);
            event.output_tokens = Some(output_tokens);

            let result = {
                let engine = self.pricing.lock().await;
                engine
                    .get_cost(&model, input_tokens, output_tokens, 0, 0)
                    .await
            };
            event.cost_usd = result.cost_usd;
            event.cost_confidence = result.cost_confidence;
            event.pricing_source = Some(result.pricing_source);
            event.pricing_version = Some(result.pricing_version);
        } else {
            // No usage data — record an unknown-cost event.
            event.cost_confidence = CostConfidence::Unknown;
            event.pricing_source = Some(PricingSource::Unknown);
        }

        self.buffer.lock().await.add_event(event.clone());
        event
    }

    /// Called when an LLM call errors. Records a failure `llm_call` event with
    /// the error string and `error_type` in its details. Never panics.
    pub async fn on_llm_error(&self, run_id: &str, error_type: &str, error: &str) -> CostEvent {
        let pending = self.pending.lock().await.remove(run_id);
        let (model, latency_ms) = match pending {
            Some(p) => (
                p.model,
                Some(p.start_time.elapsed().as_millis() as i64),
            ),
            None => ("unknown".to_string(), None),
        };

        let mut event = CostEvent::new(&self.task_id, EventType::LlmCall);
        event.provider = Some("langchain".to_string());
        event.model = Some(model);
        event.latency_ms = latency_ms;
        event.cost_confidence = CostConfidence::Unknown;
        event.pricing_source = Some(PricingSource::Unknown);
        event.details.insert(
            "error".to_string(),
            serde_json::Value::String(error.to_string()),
        );
        event.details.insert(
            "error_type".to_string(),
            serde_json::Value::String(error_type.to_string()),
        );

        self.buffer.lock().await.add_event(event.clone());
        event
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn fixtures() -> (Arc<Mutex<EventBuffer>>, Arc<Mutex<PricingEngine>>) {
        (
            Arc::new(Mutex::new(EventBuffer::new().expect("buffer"))),
            Arc::new(Mutex::new(PricingEngine::new())),
        )
    }

    #[tokio::test]
    async fn test_on_llm_end_records_event() {
        let (buffer, pricing) = fixtures();
        let handler = DexcostCallbackHandler::new(buffer.clone(), pricing, "task-lc");

        handler.on_llm_start("run-1", "gpt-4o").await;
        let event = handler.on_llm_end("run-1", 1000, 500).await;

        assert_eq!(event.event_type, EventType::LlmCall);
        assert_eq!(event.provider.as_deref(), Some("langchain"));
        assert_eq!(event.model.as_deref(), Some("gpt-4o"));
        assert_eq!(event.input_tokens, Some(1000));
        assert_eq!(event.output_tokens, Some(500));

        let buf = buffer.lock().await;
        assert_eq!(buf.event_count(), 1);
    }

    #[tokio::test]
    async fn test_on_llm_error_records_failure_with_error_type() {
        let (buffer, pricing) = fixtures();
        let handler = DexcostCallbackHandler::new(buffer.clone(), pricing, "task-lc");

        handler.on_llm_start("run-err", "gpt-4o").await;
        let event = handler
            .on_llm_error("run-err", "rate_limit", "429 Too Many Requests")
            .await;

        assert_eq!(event.event_type, EventType::LlmCall);
        assert_eq!(event.cost_confidence, CostConfidence::Unknown);
        assert_eq!(
            event.details.get("error_type"),
            Some(&serde_json::Value::String("rate_limit".to_string()))
        );
        assert_eq!(
            event.details.get("error"),
            Some(&serde_json::Value::String(
                "429 Too Many Requests".to_string()
            ))
        );

        let buf = buffer.lock().await;
        assert_eq!(buf.event_count(), 1);
    }

    #[tokio::test]
    async fn test_on_llm_end_without_start_uses_unknown_model() {
        let (buffer, pricing) = fixtures();
        let handler = DexcostCallbackHandler::new(buffer, pricing, "task-lc");

        // No matching on_llm_start — model falls back to "unknown".
        let event = handler.on_llm_end("missing-run", 0, 0).await;
        assert_eq!(event.model.as_deref(), Some("unknown"));
        assert_eq!(event.cost_confidence, CostConfidence::Unknown);
    }
}
